Adding sliding curves for DVGeoMulti

bernardopacini opened the pull request:
## Purpose

There are cases when using the intersection method in which a curve on a surface should be preserved to make sure that the patch does not change size or the curve is maintained to a certain shape. For example, maintaining the interface line between a propeller hub and nacelle is important to ensure a realistic boundary condition on each patch. This PR adds this sort of tracking.

These changes are actually made by @anilyil.

## Expected time until merged
2 weeks.

## Type of change

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing
I have run this locally and also checked the derivatives to make sure the code is correct.

## Checklist

- [X] I have run `flake8` and `black` to make sure the Python code adheres to PEP-8 and is consistently formatted
- [ ] I have formatted the Fortran code with `fprettify` or C/C++ code with `clang-format` as applicable
- [ ] I have run unit and regression tests which pass locally with my changes
- [ ] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

bernardopacini commented:
I added a test for the sliding curve feature as discussed offline with @hajdik and @sseraj. It checks to ensure that a few points remain on the lines on which they were defined, but allowed to vary along them (@anilyil let me know if it is okay with you too).

sseraj commented on tests/reg_tests/test_DVGeometryMulti.py:
```diff
@@ -315,6 +315,157 @@ def twist(val, geo, nRefAxPts=nRefAxPts):
+        DVGeoBox2 = DVGeometry(ffdFiles[1])
+
+        # Set up real DVGeometryMulti object
+        DVGeo = DVGeometryMulti(comm=comm, debug=True)
```
`debug` should be False so that running the test doesn't generate `.plt` files

sseraj commented on tests/reg_tests/test_DVGeometryMulti.py:
```diff
@@ -315,6 +315,157 @@ def twist(val, geo, nRefAxPts=nRefAxPts):
+        pts_dtype = localPts
+        DVGeo.addPointSet(pts_dtype, ptSetName, comm=comm, applyIC=True)
+
+        # Apply twist to the two intersecting boxes
```
Only one box is twisted

sseraj commented on pygeo/parameterization/DVGeoMulti.py:
```diff
@@ -2892,7 +2904,10 @@ def _getIntersectionSeam(self, comm, firstCall=False):
-                remeshedCurveConn = np.vstack((remeshedCurveConn, newBarsConn))
+                remeshedCurveConnFull = np.vstack((remeshedCurveConnFull, newBarsConn))
+
+                if curveName not in self.slidingCurves:
```
By excluding sliding curves here, they are not used as seeds for the curved-based deformation. This means that points on these curves get warped like any other point. We also still want the 'full' connectivity because that is used for projections. Is this correct? A comment or two about this would be helpful

bernardopacini replied:
I will defer to @anilyil on this.

sseraj commented on pygeo/parameterization/DVGeoMulti.py:
```diff
@@ -1195,6 +1202,10 @@ def __init__(
         self.incCurves = includeCurves
 
+        # list of curves that allow nodes to slide on them. we only use these for the projection step,
+        # but they are not included in the first line based IDWarp
```
What does "first" here mean? Also, would it be more accurate to say these curves are not included as seeds in the curve-based deformation?

bernardopacini replied:
Again, I will defer to @anilyil on this.

sseraj commented on pygeo/parameterization/DVGeoMulti.py:
```diff
@@ -184,6 +184,7 @@ def addIntersection(
         project=False,
         marchDir=1,
         includeCurves=False,
+        slidingCurves=[],
```
To avoid mutable defaults, this should be initialized to None and then set to empty in `__init__` like `featureCurves`

sseraj reviewed (commented):
This is a great feature. Thanks for adding the test as well. I have a few minor comments

bernardopacini commented:
Thanks for the feedback @sseraj! I updated the files with your comments and defer to @anilyil on the implementation-specific comments.

sseraj commented on tests/reg_tests/test_DVGeometryMulti.py:
```diff
@@ -315,6 +315,157 @@ def twist(val, geo, nRefAxPts=nRefAxPts):
+        for i in range(np.size(pts, 0)):
+            self.assertAlmostEqual(pts[i, 0], ptsUpdated[i, 0])
+            self.assertNotEqual(pts[i, 1], ptsUpdated[i, 1])
+            self.assertAlmostEqual(pts[i, 2], ptsUpdated[i, 2])
```
I was thinking on something like this:
```
# Test that the X and Z coordinates are unchanged and Y coordinates are changed
np.testing.assert_allclose(pts[:, 0], ptsUpdated[:, 0], rtol=, atol=)
np.testing.assert_allclose(pts[:, 2], ptsUpdated[:, 2], rtol=, atol=)

with self.assertRaises(AssertionError):
    np.testing.assert_allclose(pts[:, 1], ptsUpdated[:, 1], rtol=, atol=)
```
The tolerances should be tight for the first two and loose for the last one

bernardopacini commented:
> @sseraj, I addressed both of your comments. Also, thank you for spelling out pretty much all of the changes needed in the comments. Both things you said are correct. Please review my changes and mark the conversations as resolved.
> 
> @bernardopacini, this PR is good for another round of reviews, correct?

I just updated the assertions as recommended by @sseraj, now should be good to go.

sseraj commented on tests/reg_tests/test_DVGeometryMulti.py:
```diff
@@ -315,6 +315,139 @@ def twist(val, geo, nRefAxPts=nRefAxPts):
+
+        # Test that the X and Z coordinates are unchanged and Y coordinates are changed
+        np.testing.assert_allclose(pts[:, 0], ptsUpdated[:, 0], rtol=1e-4, atol=1e-10)
+        np.testing.assert_allclose(pts[:, 2], ptsUpdated[:, 2], rtol=1e-4, atol=1e-10)
```
The `rtol` for these should be tighter than 1e-4, right? I would expect the coordinates to match to 8 digits at least

bernardopacini replied:
I'll make it tighter. I matched the other tests for consistency but I'll change it, you are right that it should match essentially to numerical precision.

sseraj commented on tests/reg_tests/test_DVGeometryMulti.py:
```diff
@@ -315,6 +315,139 @@ def twist(val, geo, nRefAxPts=nRefAxPts):
+        np.testing.assert_allclose(pts[:, 2], ptsUpdated[:, 2], rtol=1e-4, atol=1e-10)
+
+        with self.assertRaises(AssertionError):
+            np.testing.assert_allclose(pts[:, 1], ptsUpdated[:, 1], rtol=1e-4, atol=1e-10)
```
This `rtol` is fine because the points should differ by more than 1e-4
